worker/worker.go

```diff
@@ -0,0 +1,239 @@
+package worker
+
+import (
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/WatchBeam/redutil/heartbeat"
+	"github.com/WatchBeam/redutil/queue"
+	"github.com/garyburd/redigo/redis"
+)
+
+// Workers are capable of fetching items from a central source in Redis and
+// processing them in a queue<->worker pattern. Items are returned from the
+// Queue as Tasks and processed by the caller. They can be either marked as
+// SUCCESSFUL or FAILED depending on processing, and will be reassigned
+// appropriately.
+type Worker struct {
+	// id is a unique ID associated with a particular worker
+	id string
+	// queue is the ByteQueue which is used to read data from the main queue
+	// in Redis.
+	queue queue.ByteQueue
+	// heart is the heartbeat.Heart which is responsible for signaling that
+	// a particular worker is alive.
+	heart heartbeat.Heart
+
+	// tasks is an internal channel used to pass tasks coming out of the
+	// queue into the caller's hands
+	tasks chan *Task
+
+	// lmu is a sync.RWMutex that protects the `lifecycle` variable.
+	lmu sync.RWMutex
+	// lifecycle is the Lifecycle used to handle the state of new tasks
+	// created from out of the queue. See the documentation on Lifecycle's
+	// interface declaration for more information on how this is used.
+	lifecycle Lifecycle
+
+	// wg is a sync.WaitGroup used to wait on when either a call to Stop()
+	// or Halt() is given.
+	wg sync.WaitGroup
+	// closer is a channel used to pass the "close" signal from the Close()
+	// function into the `recv` function internally. Since only one
+	// goroutine runs with the `recv` function for any given instance of a
+	// Worker, only one `struct{}{}` need be passed in.
+	closer chan struct{}
+	// halter is a channel used to pass the "halt" signal to all task
+	// monitoring goroutines. Since "N" goroutines may be running at any
+	// given time, N items are passed into this channel.
+	halter chan struct{}
+	// errs is an internal channel used to pass errors up to the caller.
+	errs chan error
+	// taskCount keeps a running count of the number of tasks currently
+	// being worked on so the appropriate number of messages can be passed
+	// into the `halter` channel upon calling `Halt()`. Since this number
+	// can potenntially change very quickly, it is updated using the
+	// `sync.Atomic` package.
+	taskCount uint64
```

Comment:
If you're just using a counter here, make sure to prevent tasks from being "resolved" multiple times and messing things up, That could cause the uint to underflow and Close() to block indefinitely.